src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigClientImplTest.java

```diff
@@ -277,12 +305,399 @@ public void testGetTemplateErrorWithRcError() {
+    }
+  }
+
+  // Test publishTemplate
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testPublishTemplateWithNullTemplate() throws Exception {
+    client.publishTemplate(null, false, false);
+  }
+
+  @Test
+  public void testPublishTemplateWithValidTemplate() throws Exception {
+    response.addHeader("etag", TEST_ETAG);
+    response.setContent(MOCK_TEMPLATE_RESPONSE);
+
+    Template publishedTemplate = client.publishTemplate(EXPECTED_TEMPLATE, false, false);
+
+    assertEquals(TEST_ETAG, publishedTemplate.getETag());
+    assertEquals(EXPECTED_TEMPLATE, publishedTemplate);
+    assertEquals(1605423446000L, publishedTemplate.getVersion().getUpdateTime());
+    checkPutRequestHeader(interceptor.getLastRequest());
+  }
+
+  @Test
+  public void testPublishTemplateWithValidTemplateAndForceTrue() throws Exception {
+    response.addHeader("etag", TEST_ETAG);
+    response.setContent(MOCK_TEMPLATE_RESPONSE);
+
+    Template publishedTemplate = client.publishTemplate(EXPECTED_TEMPLATE, false, true);
+
+    assertEquals(TEST_ETAG, publishedTemplate.getETag());
+    assertEquals(EXPECTED_TEMPLATE, publishedTemplate);
+    assertEquals(1605423446000L, publishedTemplate.getVersion().getUpdateTime());
+    checkPutRequestHeader(interceptor.getLastRequest(), "", "*");
+  }
+
+  @Test
+  public void testPublishTemplateWithValidTemplateAndValidateOnlyTrue() throws Exception {
+    response.addHeader("etag", TEST_ETAG);
+    response.setContent(MOCK_TEMPLATE_RESPONSE);
+    Template expectedTemplate = new Template()
+            .setETag("etag-123456789012-45")
+            .setParameters(EXPECTED_PARAMETERS)
+            .setConditions(EXPECTED_CONDITIONS)
+            .setParameterGroups(EXPECTED_PARAMETER_GROUPS)
+            .setVersion(EXPECTED_VERSION);
+
+    Template validatedTemplate = client.publishTemplate(expectedTemplate, true, false);
+
+    // check if the etag matches the input template's etag and not the etag from the server response
+    assertNotEquals(TEST_ETAG, validatedTemplate.getETag());
+    assertEquals("etag-123456789012-45", validatedTemplate.getETag());
+    assertEquals(expectedTemplate, validatedTemplate);
+    checkPutRequestHeader(interceptor.getLastRequest(), "?validateOnly=true",
+            "etag-123456789012-45");
+  }
+
+  @Test
+  public void testPublishTemplateWithEmptyTemplateResponse() throws Exception {
+    response.addHeader("etag", TEST_ETAG);
+    response.setContent("{}");
+
+    Template template = client.publishTemplate(new Template().setETag(TEST_ETAG), false, false);
+
+    assertEquals(TEST_ETAG, template.getETag());
+    assertEquals(0, template.getParameters().size());
+    assertEquals(0, template.getConditions().size());
+    assertEquals(0, template.getParameterGroups().size());
+    assertNull(template.getVersion());
+    checkPutRequestHeader(interceptor.getLastRequest());
+  }
+
+  @Test(expected = IllegalStateException.class)
```

Comment:
Here too. Which lines are supposed to throw? The test case will terminate upon encountering the first exception.

Reply:
Ah good catch! It should throw on `client.publishTemplate(new Template(), false, false);`. Separated into two tests.